src/s3fs.cpp

```diff
@@ -2708,6 +2721,472 @@ static int remote_mountpath_exists(const char* path)
+  if(!list || size < total){
+    free_xattrs(xattrs);
+    return -ERANGE;
+  }
+
+  // copy to list
+  char* setpos = list;
+  for(xattrs_t::const_iterator iter = xattrs.begin(); iter != xattrs.end(); ++iter){
+    if(0 < iter->first.length()){
+      strcpy(setpos, iter->first.c_str());
+      setpos = &setpos[strlen(setpos) + 1];
+    }
+  }
+  free_xattrs(xattrs);
+
+  return total;
+}
+
+static int s3fs_removexattr(const char* path, const char* name)
+{
+  FPRN("[path=%s][name=%s]", path, name);
+
+  if(!path || !name){
+    return -EIO;
+  }
+
+  int         result;
+  string      strpath;
+  string      newpath;
+  string      nowcache;
+  headers_t   meta;
+  xattrs_t    xattrs;
+  struct stat stbuf;
+  int         nDirType = DIRTYPE_UNKNOWN;
+
+  if(0 == strcmp(path, "/")){
+    DPRNNN("Could not change mode for mount point.");
+    return -EIO;
+  }
+  if(0 != (result = check_parent_object_access(path, X_OK))){
+    return result;
+  }
+  if(0 != (result = check_object_owner(path, &stbuf))){
+    return result;
+  }
+
+  if(S_ISDIR(stbuf.st_mode)){
+    result = chk_dir_object_type(path, newpath, strpath, nowcache, &meta, &nDirType);
+  }else{
+    strpath  = path;
+    nowcache = strpath;
+    result   = get_object_attribute(strpath.c_str(), NULL, &meta);
+  }
+  if(0 != result){
+    return result;
+  }
+
+  // get xattrs
+  if(meta.end() == meta.find("x-amz-meta-xattr")){
+    // object does not have xattrs
+    return -ENOATTR;
+  }
+  string strxattrs = meta["x-amz-meta-xattr"];
+
+  parse_xattrs(strxattrs, xattrs);
+
+  // check name xattrs
+  string strname = name;
+  if(xattrs.end() == xattrs.find(strname)){
+    free_xattrs(xattrs);
+    return -ENOATTR;
+  }
+
+  // make new header_t after deleting name xattr
+  if(xattrs[strname]){
+    delete xattrs[strname];
+  }
+  xattrs.erase(strname);
+
+  // build new xattr
+  if(0 < xattrs.size()){
+    meta["x-amz-meta-xattr"] = build_xattrs(xattrs);
+  }else{
+    meta.erase("x-amz-meta-xattr");
+  }
+
+  if(S_ISDIR(stbuf.st_mode) && IS_REPLACEDIR(nDirType)){
+    // Should rebuild directory object(except new type)
+    // Need to remove old dir("dir" etc) and make new dir("dir/")
+
+    // At first, remove directory old object
+    if(IS_RMTYPEDIR(nDirType)){
+      S3fsCurl s3fscurl;
+      if(0 != (result = s3fscurl.DeleteRequest(strpath.c_str()))){
+        free_xattrs(xattrs);
+        return result;
+      }
+    }
+    StatCache::getStatCacheData()->DelStat(nowcache);
+
+    // Make new directory object("dir/")
+    if(0 != (result = create_directory_object(newpath.c_str(), stbuf.st_mode, stbuf.st_mtime, stbuf.st_uid, stbuf.st_gid))){
+      free_xattrs(xattrs);
+      return result;
+    }
+
+    // need to set xattr header for directory.
+    strpath  = newpath;
+    nowcache = strpath;
+  }
+
+  // set xattr all object
+  meta["x-amz-copy-source"]        = urlEncode(service_path + bucket + get_realpath(strpath.c_str()));
+  meta["x-amz-metadata-directive"] = "REPLACE";
```

Comment:
Call `meta.insert("x-amz-metadata-directive", "REPLACE")` to avoid an unneeded default constructor.